pageserver/src/walredo.rs

```diff
@@ -587,419 +583,1506 @@ impl<C: CommandExt> CloseFileDescriptors for C {
-            }
-
-            // If we have more data to write and 'stdin' is writeable, do write.
-            if nwrite < writebuf.len() {
-                let in_revents = pollfds[2].revents().unwrap();
-                if in_revents & (PollFlags::POLLERR | PollFlags::POLLOUT) != PollFlags::empty() {
-                    nwrite += self.stdin.write(&writebuf[nwrite..])?;
-                } else if in_revents.contains(PollFlags::POLLHUP) {
-                    // We still have more data to write, but the process closed the pipe.
-                    return Err(Error::new(
-                        ErrorKind::BrokenPipe,
-                        "WAL redo process closed its stdin unexpectedly",
-                    ));
+            }
+            Err(io_or_timeout) => {
+                // TODO: could we somehow manage to keep the request in case we need to
+                // restart the process? see https://github.com/neondatabase/neon/issues/1700
+                // we could send it back to the queue at least here, but not so easily when it's
+                // inflight.
+
+                if let Err(e) =
+                    response.send(Err(io_or_timeout).context("Failed to write request to wal-redo"))
+                {
+                    let e = e.into_inner().expect_err("just created the value");
+                    // log this here, even though we could return an error
+                    warn!("stopping due to io or timeout error: {e:#}");
                 }
+                // we can still continue processing pipelined requests, if any. the
+                // stdout task will exit upon seeing we've dropped the result_txs.
+                return Ok(());
             }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum StdoutTaskError {
+    ReadFailed(std::io::Error),
+    StdoutClosed,
+    ReadTimeout,
+}
 
-            // If we have some data in stdout, read it to the result buffer.
-            let out_revents = pollfds[0].revents().unwrap();
-            if out_revents & (PollFlags::POLLERR | PollFlags::POLLIN) != PollFlags::empty() {
-                nresult += self.stdout.read(&mut resultbuf[nresult..])?;
-            } else if out_revents.contains(PollFlags::POLLHUP) {
-                return Err(Error::new(
-                    ErrorKind::BrokenPipe,
-                    "WAL redo process closed its stdout unexpectedly",
-                ));
+/// Always eventually fails, returning a developer friendly reason.
+#[instrument(name = "stdout", skip_all)]
+async fn stdout_task<AR>(
+    mut stdout: AR,
+    mut inter_rx: tokio::sync::mpsc::Receiver<IntermediatePayload>,
+) -> Result<(), StdoutTaskError>
+where
+    AR: tokio::io::AsyncRead + Unpin,
+{
+    use tokio::io::AsyncReadExt;
+
+    async fn read_page<AR>(mut stdout: AR, buffer: &mut Vec<u8>) -> Result<(), StdoutTaskError>
+    where
+        AR: tokio::io::AsyncRead + Unpin,
+    {
+        loop {
+            let read = stdout
+                .read_buf(buffer)
```

Comment:
The docs are a bit vague for `read_buf`, implementation is quite clear "one read". My understanding is that it will go to sleep until there's a readyness event, then try to read the remaining vec. Because of how the response protocol goes, it is that one 8192 (or more, if unlucky/stalled) byte read.

To be honest I haven't tested any interesting configurations for this, for example 7kB pipe size, but I can't see how it could go wrong.